daisy_workflows/image_build/enterprise_linux/export_metadata.py

```diff
@@ -0,0 +1,108 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Export image metadata on a GCE VM.
+
+Parameters (retrieved from instance metadata):
+google_cloud_repo: The package repo to use. Can be stable (default), staging,
+  or unstable.
+el_release: rhel6, rhel7, centos6, centos7, oraclelinux6, or oraclelinux7
+image_dest: The Cloud Storage destination for the resultant image.
+destination_prefix: The CS path to export image metadata to
+"""
+
+import json
+import logging
+import os
+import time
+import uuid
+
+import utils
+
+
+def main():
+  # Get Parameters.
+  repo = utils.GetMetadataAttribute('google_cloud_repo',
+                                    raise_on_not_found=True).strip()
+  el_release_name = utils.GetMetadataAttribute('el_release',
+                                               raise_on_not_found=True)
+  build_date = utils.GetMetadataAttribute('build_date',
+                                          raise_on_not_found=True)
+  destination_prefix = utils.GetMetadataAttribute('destination_prefix',
+                                                  raise_on_not_found=True)
+  uefi = utils.GetMetadataAttribute('rhel_uefi') == 'true'
+
+  # Mount the installer disk.
+  if uefi:
+    utils.Execute(['mount', '/dev/sdb2', '/mnt'])
+  else:
+    utils.Execute(['mount', '/dev/sdb1', '/mnt'])
+
+  logging.info('Installer root: %s' % os.listdir('/mnt'))
+
+  # Create and upload metadata of the image and packages
+  logging.info('Creating image metadata.')
+  release_time = time.time()
+  image = {
+      "id": uuid.uuid4(),
+      "family": el_release_name,
+      "name": el_release_name + build_date,
+      "version": build_date,
+      "location": "gs://gce-image-archive/centos-uefi/centos-7-v${build_date}",
+      "release_date": build_date,
+      "release_time": release_time,
+      "state": "Active",
+      "environment": "prod",
+      "packages": []
+  }
+
+  # Read list of guest package
+  with open("guest_package") as f:
+    guest_packages = f.read().splitlines()
+
+  for package in guest_packages:
+    # The URL is a github repo link which don't contain commit hash
+    cmd = "rpm -q --queryformat='%{NAME} %{VERSION} %{RELEASE} %{URL}\n'" \
+          + package
+    code, stdout = utils.Excute(cmd, capture_output=True)
+    if code == 0:
+      splits = stdout.decode('utf-8').split('\t\b')
+      package_name = splits[0]
+      package_version = splits[1] + "-" + splits[2]
+      package_release_time = package_version[0:package_version.index(".")]
+      metadata = {
+          "id": uuid.uuid4(),
+          "name": package_name,
+          "version": package_version,
+          "commit_hash": "",
+          "release_date": package_release_time,
+          "stage": repo
+      }
+      image["packages"].append(metadata)
+
+      with open('/tmp/metadata.json', 'w') as f:
+        f.write(json.dumps(image))
```

Comment:
uh, this will overwrite each time because mode `w` is overwriting. if you switch this to `a`, you'll append, but then you'll be appending full json blobs, which produces an invalid json blob.

i think you need to build up a python dict with all values and then write once at the end.

Reply:
i think the mistake is "indent". the "package" metadata is a array field in the "image". 
json.dump(image) should produce 👍 
`{
  "id": "image-uuid",
  "family": "debian-8",
  "name": "gold",
  "release_date": "release_date",
  "state": "state",
  "environment" : "environment",
  "packages" : [
    {
      "id": "package-1-uuid",
      "name": "package_name",
      "version": "package_version",
      "commit_hash": "package_hash_1",
      "release_date": "package_release_date",
      "stage": "repo"
    },
    {
      "id": "package-2-uuid",
      "name": "package_name-2",
      "version": "package_version-2",
      "commit_hash": "package_hash_2",
      "release_date": "package_release_date-2",
      "stage": "repo"
    } 
   ]
}`